prow pod-utils/clone: make git checkouts more reproducible

ixdy opened the pull request:
This PR makes two changes in an aim to make builds more reproducible. It's modeled after #7480, which implemented this functionality for `jenkins/bootstrap.py`. It is part of addressing #7760

- pod-utils/clone: set `GIT_AUTHOR_DATE` and `GIT_COMMITTER_DATE` on all merge commits, using sequential seconds after the timestamp of the base ref. This way, if the same set of refs are used on multiple jobs, they'll all get the same commit timestamps and git digests. If we can't read the timestamp of the base ref for some reason, we'll just use sequential seconds starting from now.
- <strike>entrypoint: set `SOURCE_DATE_EPOCH` to the timestamp of HEAD. This is basically what https://reproducible-builds.org/docs/source-date-epoch/ recommends, and means that rebuilding at the same ref will use the same timestamp in build systems that support it (such as bazel and the kubernetes make-based build). This isn't set if we can't read it (e.g. no repo was checked out).</strike>

I'm still trying to figure out how to adequately test this new functionality, hence WIP. If you have suggestions on where to add tests, that'd be welcome.

I've manually tested prow/cmd/clonerefs, and it seems to be doing the right thing with `GIT_AUTHOR_DATE` and `GIT_COMMITTTER_DATE`.

/hold

@stevekuznetsov @BenTheElder @fejta

ixdy commented:
entrypoint seems to work, too.

From my test-infra dir:
```console
$ bazel-bin/prow/cmd/entrypoint/linux_amd64_pure_stripped/entrypoint --process-log=/tmp/log --marker-file=/tmp/marker -- bash -c 'echo SOURCE_DATE_EPOCH is ${SOURCE_DATE_EPOCH}'
SOURCE_DATE_EPOCH is 1544138570
```
From a random non-git dir, or if I make `git` print out garbage:
```console
$ .../entrypoint --process-log=/tmp/log --marker-file=/tmp/marker -- bash -c 'echo SOURCE_DATE_EPOCH is ${SOURCE_DATE_EPOCH}'
SOURCE_DATE_EPOCH is
```

ixdy commented on prow/pod-utils/clone/clone_test.go:
```diff
@@ -315,18 +321,37 @@ func TestCommandsForRefs(t *testing.T) {
-		if actual, expected := commandsForRefs(testCase.refs, testCase.dir, testCase.gitUserName, testCase.gitUserEmail, testCase.cookiePath, testCase.env), testCase.expected; !reflect.DeepEqual(actual, expected) {
-			t.Errorf("%s: generated incorrect commands: %v", testCase.name, diff.ObjectGoPrintDiff(expected, actual))
+		actual := commandsForRefs(testCase.refs, testCase.dir, testCase.gitUserName, testCase.gitUserEmail, testCase.cookiePath, testCase.env)
+		// reflect.DeepEqual always treats functions as unequal,
```
I'm not really sure how to refactor this while preserving the test coverage provided by `clone_test.go` without fully doubling down and creating fake git repos to test against.

(I'm also not super excited to refactor this code right now.)

ixdy replied:
indeed, the current indirectness was added to facilitate testing: 33f1fd40088f09c3defdb488cb20e0448e3bfeb4

I'll admit that I'm not really sure how to test the timestamp behavior without creating a full integration test that actually creates a fake git repo and then checks out a set of refs.

stevekuznetsov replied:
This could be clearer if we made it less generic and explicitly supported the two env vars and explicitly for clones. Unclear if we need that.

ixdy commented:
I've also added some tests for the entrypoint functionality, since that's a bit more self-contained.

ixdy commented:
/hold cancel

mithrav commented:
First off, Kudos for the detailed PR description!

krzyzacy commented on prow/pod-utils/clone/clone.go:
```diff
@@ -101,6 +103,29 @@ func commandsForRefs(refs kube.Refs, dir, gitUserName, gitUserEmail, cookiePath
+	// value in the GIT_AUTHOR_DATE and GIT_COMMITER_DATE environment variables.
+	fakeTimestamp := int(time.Now().Unix())
+	if len(refs.Pulls) > 0 {
+		gitShowCommand := gitCommand("show", "-s", "--format=format:%ct", refs.BaseRef)
```
assume here you want to get the Pull ref instead of BaseRef?

krzyzacy commented:
Compared with the original bootstrap PR and looks fine on what it want to achieve.

Maybe it's easier to execute the commands right away instead of queue them up, and we don't need to explicitly handle the before and after side effects in all the commands?

stevekuznetsov commented on prow/entrypoint/run_test.go:
```diff
@@ -124,3 +170,101 @@ func compareFileContents(name, file, expected string, t *testing.T) {
 	}
 }
+
+func TestGitHeadTimestamp(t *testing.T) {
```
This adds a dependency on `git` to the test image environment  -- do we need to update the image for the test job?

ixdy replied:
doesn't seem like it? I think the golang image contains git already, since it's needed for things like `go get`.

ixdy commented:
I've refactored this a bit to break up the lists of commands into two functions, one for all the commands to initialize the repo / check out the base ref, and then another for pull refs (which does the environment variable munging). The list of commands are still useful for at least two reasons:
a) unit tests
b) logging/error handling/aborting if any command fails without having to write `err != nil` 20 times

One consequence of this refactoring is that I have basically the same `gitHeadTimestamp` function in two places now, one in `entrypoint`, one in `pod-utils/clone`. I'm not sure where would be a good common place for that to live.

stevekuznetsov commented on prow/entrypoint/run.go:
```diff
@@ -206,3 +211,20 @@ func gracefullyTerminate(command *exec.Cmd, done <-chan error, gracePeriod time.
+// UNIX epoch. If unable to read the timestamp for any reason (such as missing
+// the git, or not using a git repo), returns 0 and an error.
+func gitHeadTimestamp() (int, error) {
+	gitOutput, err := exec.Command("git", "show", "-s", "--format=format:%ct", "HEAD").Output()
```
Why do we need this on the entrypoint? This also assumes the test image will have `git` which is not the case today

ixdy replied:
this is basically just following https://reproducible-builds.org/docs/source-date-epoch/#git-repository - we want to set `SOURCE_DATE_EPOCH` to the timestamp of the last modification.

I hadn't thought about the test image not having `git`. In kubernetes we would pretty much require it to get a valid version string, but I guess that's not the case everywhere. We could instead save the git timestamp (and maybe other info?) into a file somewhere? Not sure where though.

ixdy replied:
(also, this fails silently - if `git` isn't available, we'll just not set `SOURCE_DATE_EPOCH`, which is the existing behavior.)

stevekuznetsov replied:
Yeah pod-utils doesn't want to provide any restriction on the image used. Can we thread through the data using file passing in shared volumes? You can see how we do this for other data in decorate.go

ixdy replied:
I guess you are making a reasonable argument that this doesn't belong in `entrypoint`.

Alternatives I see:
a) Add it to the bootstrap image, which would cover the majority of kubernetes' needs (and would be easy to implement). Anyone not using a bootstrap-derived image would need to implement this themselves.
b) Make clonerefs save out metadata to a separate file (either just the timestamp, or maybe additionally other git info like `git describe`?) We could then either plumb this file through or pass it via an environment variable or both. I'm not really sure what would be best. This would be quite a bit more work, and I'm not sure how I'd even test it, since we don't really have any integration tests for the pod-utils components.

At this point I'm leaning towards a), since it unblocks our migration to pod-utils. If anyone wants a more general implementation, they're welcome to work on it.

stevekuznetsov replied:
I think `b)` would be a good approach for the future -- I'm thinking something like adding a flag to point to the location where to write the data, then throwing down a new volume shared between clonerefs and entrypoint, and similarly plumbing through a flag to entrypoint to consume it.

ixdy commented:
I've reverted all changes to `entrypoint` except one small additional test case which seems useful to keep.

I moved the git-based tests to `clone_test`, and will likely follow up with bootstrap image changes in another PR.

PTAL?

stevekuznetsov commented:
I think this makes sense to me. @fejta @cjwagner did you want to review?

/lgtm
/hold

fejta commented on prow/entrypoint/run_test.go:
```diff
@@ -77,6 +77,13 @@ func TestOptions_Run(t *testing.T) {
+		{
+			// Ensure that environment variables get passed through
+			name:           "$PATH is set",
+			args:           []string{"sh", "-c", "echo -n $PATH"},
```
This test fails on mac

fejta replied:
linux:
```console

$ sh -c "echo -n $PATH"
this:that:other$
```

mac:
```console
$ sh -c "echo -n $PATH"
-n this:that:other
$
```

🙄

fejta replied:
https://github.com/kubernetes/test-infra/pull/10716
